_posts/2019-09-30-introducing-arrow-flight.md

```diff
@@ -0,0 +1,260 @@
+the endpoints must be consumed.
+
+This multiple-endpoint pattern has a number of benefits:
+
+* Endpoints can be read by clients in parallel
+* The service that serves the `GetFlightInfo` "query planning" request can
+  delegate work to sibling services to take advantage of data locality or
+  simply to help with load balancing
+* Nodes in a distributed cluster can take on different roles. For example, a
+  subset of nodes might be responsible for planning queries while other nodes
+  exclusively fulfill data stream ("DoGet") requests
+
+Here is an example diagram of a multi-node architecture with split service
+roles:
+
+<div align="center">
+<img src="{{ site.baseurl }}/img/20191014_flight_complex.png"
+     alt="Flight Complex Architecture"
+     width="60%" class="img-responsive">
+</div>
+
+## Actions: Extending Flight with application business logic
+
+While the `GetFlightInfo` request supports sending opaque serialized commands
+when requesting a dataset, a client may need to be able to ask a server to
+perform other kinds of operations. For example, a client may request for a
+particular dataset to be "pinned" in memory so that subsequent requests from
+other clients are served faster.
+
+A Flight service can thus optionally define "actions" which are carried out by
+the `DoAction` RPC. An action request contains the name of the action being
+performed and optional serialized data containing further needed
+information. The result of an action is a gRPC stream of opaque binary results.
+
+An example action would be the command `'ListDatasets'` which could return a
+stream of dataset names that are available on that server.
+
+Note that it is not required for a server to implement any actions, and actions
+need not return results.
+
+## Encryption and Authentication
+
+Flight supports encryption out of the box using gRPC's built in TLS / OpenSSL
+capabilities.
+
+For authentication, there are extensible authentication handlers for the client
+and server that permit simple authentication schemes (like user and password)
+as well as more involved authentication such as Kerberos. The Flight protocol
+comes with a built-in `BasicAuth` so that user/password authentication out of
+the box without custom development.
+
+## Middleware and Tracing
+
+gRPC has the concept of "interceptors" which have allowed us to develop
+developer-defined "middleware" that can provide instrumentation of or telemetry
+for incoming and outgoing requests. One such framework for such instrumentation
+is the [OpenTracing][6] framework
+
+## gRPC, but not only gRPC
+
+We specify server locations for `DoGet` requests using RFC 3986 compliant
+URIs. For example, TLS-secured gRPC may be specified like
+`grpc+tls://$HOST:$PORT`.
+
+While we think that using gRPC for the "command" layer of Flight servers makes
+sense, we may wish to support transport layers other than gRPC for data
+transfer. One example is [RDMA][7].
```

Comment:
Is RDMA a layer at the same level as gRPC? It would seem it's more at the Ethernet (or perhaps IP) level. i.e. you still need some RPC protocol over that.

Reply:
Yes, it's definitely a bit hand-wavy. I added some language to indicate that it would be a design and development effort to make this possible